### PR TITLE
Add sexpr types

### PR DESCRIPTION
**Pull request by manuranga**
Add recursive sexpr types
Refactor using match stmt
Part of #1031 relates to #1042

### CODE REVIEW

File: compiler/modules/types/core.bal [hunk 1]
```diff
@@ -1015,10 +1016,30 @@ public function stringSubtype(SemType t) returns StringSubtype|boolean {
     return <boolean|StringSubtype>subtypeData(t, BT_STRING);
 }
 
+public function listSubtype(SemType t) returns Bdd {
```
Comment (jclark) [inline compiler/modules/types/core.bal:1018]: Order: list, mapping, table, xml, error, function

File: compiler/modules/types/error.bal [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // Implementation specific to basic type error.
 
-public function errorDetail(Context cx, SemType detail) returns SemType {
+public function errorDetail(SemType detail) returns SemType {
```
Comment (jclark) [inline compiler/modules/types/error.bal:3]: Let's not do this in this PR.

File: compiler/modules/types/sexprSerialize.bal [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+|};
+type MemoTable table<Memo> key (id);
+
+type SerializationContext record {|
+    Context tc;
+    map<ts:Atom> bindings = {};
+    MemoTable memos = table[];
+|};
+
+public function sexprFormSemType(Context tc, SemType ty) returns [ts:Type, map<ts:Atom>] {
```
Comment (jclark) [inline compiler/modules/types/sexprSerialize.bal]: `From`

Comment (jclark) [reply compiler/modules/types/sexprSerialize.bal]: `map<ts:Atom>` should be an argument

Comment (jclark) [reply compiler/modules/types/sexprSerialize.bal]: toSexpr

File: compiler/modules/types/error.bal (hunk 2 above)
Comment (manuranga) [reply compiler/modules/types/error.bal:3]: 1) Currently type sexpr parser only refer to `t:Env`, never to `t:Context`.
2) `Context cx` was only added with Lochana's changes, even then it was never used.

Therefor, I propose we remove it in (or before) this PR.

**Review by jclark — APPROVED** [review]

**Pull request merged**